File: Python/329.longest-increasing-path-in-a-matrix.py
#
# @lc app=leetcode id=329 lang=python
#
# [329] Longest Increasing Path in a Matrix
#
# https://leetcode.com/problems/longest-increasing-path-in-a-matrix/description/
#
# algorithms
# Hard (41.37%)
# Likes:    1325
# Dislikes: 22
# Total Accepted:    110.3K
# Total Submissions: 265.8K
# Testcase Example:  '[[9,9,4],[6,6,8],[2,1,1]]'
#
# Given an integer matrix, find the length of the longest increasing path.
# 
# From each cell, you can either move to four directions: left, right, up or
# down. You may NOT move diagonally or move outside of the boundary (i.e.
# wrap-around is not allowed).
# 
# Example 1:
# 
# 
# Input: nums = 
# [
# ⁠ [9,9,4],
# ⁠ [6,6,8],
# ⁠ [2,1,1]
# ] 
# Output: 4 
# Explanation: The longest increasing path is [1, 2, 6, 9].
# 
# 
# Example 2:
# 
# 
# Input: nums = 
# [
# ⁠ [3,4,5],
# ⁠ [3,2,6],
# ⁠ [2,2,1]
# ] 
# Output: 4 
# Explanation: The longest increasing path is [3, 4, 5, 6]. Moving diagonally
# is not allowed.
# 
# 
#

# @lc code=start
class Solution(object):
    def longestIncreasingPath(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: int
        """

        dp = [[-1] * len(matrix[0]) for _ in range(len(matrix))]

        ans = 1
        for i in range(len(matrix)):
            for j in range(len(matrix[0])):
                l = self.dfs(matrix, i, j, dp)
                ans = max(ans, l)
        return ans

    def dfs(self, matrix, i, j, dp):
        if dp[i][j] != -1:
            return dp[i][j]
        max_len = 1
        for d in [[0, 1], [0, -1], [1, 0], [-1, 0]]:
            x, y = i + d[0], j + d[1]
            if x < 0 or x >= len(matrix) or y < 0 or y >= len(matrix[0]):
                continue
            if matrix[x][y] <= matrix[i][j]:
                continue
            max_len = max(max_len, self.dfs(matrix, x, y, dp) + 1)
        dp[i][j] = max_len
        return dp[i][j]


# @lc code=end

sol = Solution()
matrix = [[9,9,4],[6,6,8],[2,1,1]]
print(sol.longestIncreasingPath(matrix))

File: C++/329.longest-increasing-path-in-a-matrix.cpp
/*
 * @lc app=leetcode id=329 lang=cpp
 *
 * [329] Longest Increasing Path in a Matrix
 *
 * https://leetcode.com/problems/longest-increasing-path-in-a-matrix/description/
 *
 * algorithms
 * Hard (43.53%)
 * Likes:    2176
 * Dislikes: 41
 * Total Accepted:    160.3K
 * Total Submissions: 367.3K
 * Testcase Example:  '[[9,9,4],[6,6,8],[2,1,1]]'
 *
 * Given an integer matrix, find the length of the longest increasing path.
 * 
 * From each cell, you can either move to four directions: left, right, up or
 * down. You may NOT move diagonally or move outside of the boundary (i.e.
 * wrap-around is not allowed).
 * 
 * Example 1:
 * 
 * 
 * Input: nums = 
 * [
 * ⁠ [9,9,4],
 * ⁠ [6,6,8],
 * ⁠ [2,1,1]
 * ] 
 * Output: 4 
 * Explanation: The longest increasing path is [1, 2, 6, 9].
 * 
 * 
 * Example 2:
 * 
 * 
 * Input: nums = 
 * [
 * ⁠ [3,4,5],
 * ⁠ [3,2,6],
 * ⁠ [2,2,1]
 * ] 
 * Output: 4 
 * Explanation: The longest increasing path is [3, 4, 5, 6]. Moving diagonally
 * is not allowed.
 * 
 * 
 */
#include<vector>
using namespace std;

// @lc code=start
class Solution {
public:
    int longestIncreasingPath(vector<vector<int>>& matrix) {
        if (matrix.empty() || matrix[0].empty()) return 0;

        int  m = matrix.size(), n = matrix[0].size();
        int ans = 1;
        vector<vector<int>> dp(m, vector<int>(n, 0));
        for (int i=0; i<m; ++i) {
            for (int j=0; j<n; ++j) {
                ans = max(ans, DFS(matrix, dp, i, j));
            }
        }
        return ans;
    }
    int DFS(vector<vector<int>> &matrix, vector<vector<int>> &dp, int i, int j) {
        vector<vector<int>> dirs = {{0, -1}, {-1, 0}, {0, 1}, {1, 0}};

        if (dp[i][j]) return dp[i][j];
        
        int m = matrix.size(), n = matrix[0].size();
        
        int ans = 1;
        for (auto d : dirs) {
            int x = i + d[0], y = j + d[1];
            if (x<0 || x>=m || y<0 || y>=n || matrix[x][y] <= matrix[i][j]) 
                continue;
            int len = 1 + DFS(matrix, dp, x, y);
            ans = max(ans, len);
        }
        dp[i][j] = ans;
        return ans;
    }
};
// @lc code=end

